src/pip/_internal/cache.py

```diff
@@ -93,13 +138,19 @@ def _get_candidates(self, link, canonical_package_name):
         if not self.allowed_formats.intersection(formats):
             return []
 
-        root = self.get_path_for_link(link)
-        try:
-            return os.listdir(root)
-        except OSError as err:
-            if err.errno in {errno.ENOENT, errno.ENOTDIR}:
-                return []
-            raise
+        candidates = []
+        path = self.get_path_for_link(link)
+        if os.path.isdir(path):
+            candidates.extend(os.listdir(path))
+        # TODO remove legacy path lookup in pip>=21
```

Comment:
We should make an issue to add a deprecation warning around here when this is merged.

Reply:
What kind of deprecation warning do you have in mind?

Reply:
I'm wondering the same as I think `pip._internal.utils.deprecation.deprecated` would be too noisy for such a thing... 
Maybe something like https://github.com/pypa/pip/blob/b3aced9617478a65b06eefde54e2b8ca6db64852/src/pip/_internal/utils/deprecation.py#L101-L102 to make sure we don't forget about it ?

Reply:
To not forget about things to do in the future, you could maybe use GitHub milestones and assign such TODO issues to future milestones. Now that pip has a regular release cadence it might be easy enough to manage.